js_modules/dagit/packages/core/src/graph/asyncGraphLayout.ts

```diff
@@ -0,0 +1,163 @@
+  cacheKey: string;
+};
+
+type Action =
+  | {type: 'loading'}
+  | {type: 'layout'; payload: {layout: OpGraphLayout | AssetGraphLayout; cacheKey: string}};
+
+const reducer = (state: State, action: Action): State => {
+  switch (action.type) {
+    case 'loading':
+      return {loading: true, layout: null, cacheKey: ''};
+    case 'layout':
+      return {
+        loading: false,
+        layout: action.payload.layout,
+        cacheKey: action.payload.cacheKey,
+      };
+    default:
+      return state;
+  }
+};
+
+const initialState: State = {
+  loading: false,
+  layout: null,
+  cacheKey: '',
+};
+
+/**
+ * Todo: It would be nice to merge these hooks into one, passing the sync + async layout methods in as args.
+ * I tried but felt like the complexity wasn't worth the code savings. The key problem is that the layout
+ * functions take different args and a generic solution loses the types.
+ */
+
+export function useOpLayout(ops: ILayoutOp[], parentOp?: GraphExplorerSolidHandleFragment_solid) {
+  const [state, dispatch] = React.useReducer(reducer, initialState);
+
+  const cacheKey = _opLayoutCacheKey(ops, parentOp);
+
+  React.useEffect(() => {
+    async function runAsyncLayout() {
+      dispatch({type: 'loading'});
+      const layout = await asyncGetFullOpLayout(ops, parentOp);
+      dispatch({
+        type: 'layout',
+        payload: {layout: layout, cacheKey},
+      });
+    }
+
+    if (ops.length < ASYNC_LAYOUT_SOLID_COUNT) {
+      const layout = getFullOpLayout(ops, parentOp);
+      dispatch({type: 'layout', payload: {layout, cacheKey}});
+    } else {
+      void runAsyncLayout();
+    }
+  }, [cacheKey, ops, parentOp]);
+
+  return {
+    loading: state.loading || !state.layout || state.cacheKey !== cacheKey,
+    async: ops.length >= ASYNC_LAYOUT_SOLID_COUNT,
+    layout: state.layout as OpGraphLayout | null,
+  };
+}
+
+export function useAssetLayout(graphData: GraphData) {
+  const [state, dispatch] = React.useReducer(reducer, initialState);
+
+  const cacheKey = _assetLayoutCacheKey(graphData);
+
+  React.useEffect(() => {
+    async function runAsyncLayout() {
+      dispatch({type: 'loading'});
+      const layout = await asyncGetFullAssetLayout(graphData);
+      dispatch({
+        type: 'layout',
+        payload: {layout: layout, cacheKey},
+      });
+    }
+
+    if (Object.keys(graphData.nodes).length < ASYNC_LAYOUT_SOLID_COUNT) {
+      const layout = getFullAssetLayout(graphData);
+      dispatch({type: 'layout', payload: {layout, cacheKey}});
+    } else {
+      void runAsyncLayout();
+    }
+  }, [cacheKey, graphData]);
+
+  return {
+    loading: state.loading || !state.layout || state.cacheKey !== cacheKey,
+    async: Object.keys(graphData.nodes).length >= ASYNC_LAYOUT_SOLID_COUNT,
+    layout: state.layout as AssetGraphLayout | null,
+  };
+}
+
+export {layoutOp} from './layout';
+export type {OpGraphLayout, OpLayout, ILayout, OpLayoutEdge, IPoint} from './layout';
```

Comment:
@hellendag  I'm curious if you remember why you re-exported these types here? Were we / are we trying to avoid external imports from ./layout?

Reply:
I don't remember, unfortunately. Possibly something related to circular deps?

Reply:
Ok - I verified that the web worker is still loading a small subset of deps, etc. so I think we're set

<img width="985" alt="image" src="https://user-images.githubusercontent.com/1037212/163058140-743224cc-67b4-4d21-bab1-3f7c22d346ef.png">